File: ruby/030.rb
# Find the sum of all the numbers that can be written as the sum of fifth powers
# of their digits.

def equals_sum_of_5_powers?(n)
    n == (n.to_s.split("").map { |digit| digit.to_i**5 }.reduce(:+))
end

# The upper limit is 6*9^5 = 354294
puts (10..354294).find_all { |n| equals_sum_of_5_powers?(n) }.reduce(:+)

File: clojure/src/project_euler_clojure/030.clj
;; Find the sum of all the numbers that can be written as the sum of
;; fifth powers of their digits.

(ns project-euler-clojure.030)

(defn sum-fifth-power-digits [n]
  (->> (str n)
       (map char)
       (map #(Character/getNumericValue %))
       (map #(int (Math/pow % 5)))
       (reduce +)))

;; The upper limit is 6*9^5 = 354294
(defn problem-030 []
  (->> (range 10 354295)
       (filter #(= % (sum-fifth-power-digits %)))
       (reduce +)))
